### PR TITLE
Allow restricting OidcRequestFilters to specific OIDC endpoints

### LINKED ISSUE #37256: Allow restricting OidcRequestFilter to specific OIDC endpoints and scopes
### Description

Right now a custom OIDC Request filter will be invoked for all of OIDC endpoint requests. For example, if these requests are done by `quarkus-oidc` then it can be up to 5 endpoints, so the filter which may want to intercept a JWKS endpoint request will need to filter request paths etc.

Additionally, if `quarkus-oidc-client` is used alongside `quarkus-oidc` then the filter will also be run when `quarkus-oidc-client` makes requests to the token endpoint

### Implementation ideas

Allow each filter restrict its application to a specific OIDC endpoint and scope with 2 enums

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #37256

### CODE REVIEW

File: extensions/oidc-common/runtime/src/main/java/io/quarkus/oidc/common/OidcRequestFilter.java [hunk 1]
```diff
@@ -8,6 +8,54 @@
  * which are made from the OIDC adapter to the OIDC provider
  */
 public interface OidcRequestFilter {
+
+    enum Scope {
+        ALL,
+        /**
+         * Applies to OIDC endpoint requests made by OIDC client, independently
+         * of OIDC authorization code flow bearer token authentication mechanism
```
Comment (pedroigor) [inline extensions/oidc-common/runtime/src/main/java/io/quarkus/oidc/common/OidcRequestFilter.java:16]:
```suggestion
         * of OIDC authorization code flow and bearer token authentication mechanism
```

**Review by pedroigor — COMMENTED** [review]
A few comments ...

**Comment by sberyozkin** [comment]
@pedroigor I'll need to update the docs

**Comment by sberyozkin** [comment]
@pedroigor Let me drop `scope` property for now as it addresses a rare case which has not yet occurred, and we can review it when users ask for it

**Comment by sberyozkin** [comment]
Let me also investigate the option of injecting `OidcConfigurationMetadata`

**Comment by pedroigor** [comment]
@sberyozkin The point about injecting metadata is that we still allow users to perform decision checks based on the client request URL and the URL from the metadata. For instance, to filter only requests to the JWKS endpoint, we would have something like that in a filter implementation:

```

if (metadata.getJwksUrl().equals(request.url()) { // do filter request }

```

**Comment by sberyozkin** [comment]
Hi @pedroigor,  I've done some updates to this PR, and @michalvavrik also helped out analyzing the metadata injection options. So here is is the summary:

* I've got rid of the `Scope` enum for now as agreed, lets see if users will ever need it
* I've added a support for the filters to access `OidcConfigurationMetadata` - Michal and myself looked into it and injecting it via the request scope is messy and tricky, another option was to pass via the Vertx duplicated context, but eventually the simplest solution was found where we can simply pass it as one of the properties in `OidcRequestContextProperties` - which is available as a filter method parameter. Unfortunately, Vert.x `HttpRequest.uri()` returns relative path only while the discovered metadata contains absolute URI, so users would have to write the code like this one: https://github.com/quarkusio/quarkus/pull/37257/files#diff-8f3225844cc7a339b14050ce3a5d9e172a0c0d7628ac0f01236c357020c7d9c4R23 but I have to admit I'm not too excited about it, as one need to check the JWK/etc endpoints for null (will be null for OAuth2 servers, or when the discovery is disabled), but it is definitely a viable option in simple setups
* Retained an Endpoint enum option: https://github.com/quarkusio/quarkus/pull/37257/files#diff-3425fd162ebeddd5d267c2b1032832f28419cb5605b5bf74ccd0bfba34773820, for me it looks like the easiest option - it is not difficult to maintain - I can't imagine this enum changing, if it at all, during the next 6 or 12 months :-). Annotation based approach would definitely be trickier to implement - this will involve the addition of 7 or so new annotations to the API among other things which I'd probably like to avoid.

IMHO it is a reasonable compromise - if users want to filter all the endpoints or say only 2 of them, then `OidcConfigurationMetadata` checks can be handy, for single endpoints, `Endpoint` enum works very well IMHO.

How does it sound ?

**Comment by sberyozkin** [comment]
Dropped a noteworth feature as it is more like a necessary improvement

**Comment by sberyozkin** [comment]
/cc @pedroigor @calvernaz

Hey Pedro, I hope you'll be glad to know I've decided to follow your suggestion and I've dropped `OidcRequestFilter#endpoint()` which this PR started from, and users can restrict with annotations, for example: https://github.com/quarkusio/quarkus/pull/37257/files#diff-3425fd162ebeddd5d267c2b1032832f28419cb5605b5bf74ccd0bfba34773820R15

So, now users can either restrict to specific endpoints using annotations or use `OidcConfigurationMetadata` to filter requests to all the endpoints.

**Comment by sberyozkin** [comment]
Unfortunately I forgot again to update the docs, done now

**Review by pedroigor — APPROVED** [review]

**Pull request merged**